osm2streets/src/movements.rs

```diff
@@ -0,0 +1,187 @@
+    let mut lanes_offset = (src_lane.number() as isize) - (dst_lane.number() as isize);
+    let mut src_i: usize = 0;
+    let mut dst_i: usize = 0;
+
+    // Pair up the lanes, offsetting one side so the match lanes line up.
+    let mut result = Vec::new();
+    loop {
+        let (src_lane, src_pt) = src_lanes[src_i];
+        let (dst_lane, dst_pt) = dst_lanes[dst_i];
+        result.push(TurnMovement {
+            from: src_lane,
+            to: dst_lane,
+            path: PolyLine::new(vec![src_pt, dst_pt])
+                .unwrap_or_else(|_| PolyLine::must_new(vec![src_pt, Pt2D::new(0.0, 0.0), dst_pt])),
+        });
+
+        // Increment the appropriate road until the offset is reached...
+        if lanes_offset > 0 {
+            src_i += 1;
+            lanes_offset -= 1;
+        } else if lanes_offset < 0 {
+            dst_i += 1;
+            lanes_offset += 1;
+        }
+        // ...otherwise increment both that haven't run out of lanes.
+        else {
+            let mut advanced = false;
+            if src_i + 1 < src_len {
+                src_i += 1;
+                advanced = true;
+            }
+            if dst_i + 1 < dst_len {
+                dst_i += 1;
+                advanced = true;
+            }
+            if !advanced {
+                break;
+            }
+        }
+    }
+    Some(result)
+}
+
+/// Limitations: ignores both ways lanes
+pub fn default(
+    (src_road, src_end): (&Road, RoadEnd),
+    (dst_road, dst_end): (&Road, RoadEnd),
+    driving_side: DrivingSide,
+) -> Option<Vec<TurnMovement>> {
+    let mut result = Vec::new();
+
+    // Figure out which lanes (left-to-right) are involved in driving from src to dst.
+    let mut src_lanes = match src_end {
+        RoadEnd::End => driving_lane_endpoints(src_road, Direction::Fwd, RoadEnd::End),
+        RoadEnd::Start => driving_lane_endpoints(src_road, Direction::Back, RoadEnd::Start),
+    };
+    let mut dst_lanes = match dst_end {
+        RoadEnd::Start => driving_lane_endpoints(dst_road, Direction::Fwd, RoadEnd::Start),
+        RoadEnd::End => driving_lane_endpoints(dst_road, Direction::Back, RoadEnd::End),
+    };
+
+    if src_lanes.is_empty() || dst_lanes.is_empty() {
+        return Some(result);
+    }
+
+    // Order the lanes inside-to-outside.
+    if driving_side == DrivingSide::Left {
+        src_lanes.reverse();
+        dst_lanes.reverse();
+    }
+
+    let src_len = src_lanes.len();
+    let dst_len = dst_lanes.len();
+    let mut src_i: usize = 0;
+    let mut dst_i: usize = 0;
+
+    // Pair up the lanes, matching the inside lanes to each other.
+    loop {
+        let (src_lane, src_pt) = src_lanes[src_i];
+        let (dst_lane, dst_pt) = dst_lanes[dst_i];
+        result.push(TurnMovement {
+            from: src_lane,
+            to: dst_lane,
+            path: PolyLine::new(vec![src_pt, dst_pt])
+                .unwrap_or_else(|_| PolyLine::must_new(vec![src_pt, Pt2D::new(0.0, 0.0), dst_pt])),
```

Comment:
This is where I would want to create a nice curved path, instead of just jamming two points in.

After working with it, it might be better to remove the geometry aspect from here and calculate it later, or on the fly. That matches up better with where geometry of other kinds is calculated.

Also, the `or_else` polyline sometimes crashes when the first one does, but not always!

Reply:
On-the-fly geom could be nice, yeah, and would have reduce file size requirements maybe. (But it's not a clear-cut tradeoff; lazily calculating in a UI later can be sluggish.)

> Also, the or_else polyline sometimes crashes when the first one does, but not always!

I actually didn't look at this carefully; inserting (0, 0) in the middle is an attempt to make the total line length larger? In what kind of cases is the above `PolyLine::new(vec![src_pt, dst_pt])` failing -- when the two points are very close together? What should we do in those cases -- maybe it's valid to just have no movement geometry?

Reply:
Yes, when the points are close together. Given our current approach, we shouldn't have any zero length movements, but I think it happens coincidently sometimes.

I'm going to experiment with removing `path` from `TurnMovement` so that turns and movement are only concerned with connectivity for now (and don't have to be recalculated if trim distances change, or something like that). I want to think about storing lane centerlines and movement paths in the same way.